### ISSUE TITLE
Split package and non-GPL variant

### ISSUE THREAD

**Issue by mbargull** [conda-forge/cryptominisat-feedstock#9]
We are currently considering to make CMS the new SAT solver to power `conda` itself (ref: https://github.com/conda/conda/pull/7676#discussion_r211319757).
If this comes to fruition (which I'm rooting for), we likely want to make a smaller package out of this. This package currently includes:
1. `lib/libcryptominisat5`
2. `bin/cryptominisat5`, `bin/cryptominisat5_simple`
3. `pycryptosat`

For the usage in `conda` we wouldn't need the executables and would also want to avoid to include `bin/cryptominisat5` because it's linked to Boost and `boost-cpp` might be a bit heavy to add to Miniconda.
So possibly, we may want to split this into subpackages  `libcryptominisat`, `cryptominisat`, and `pycryptosat` or something.
Also, I assume that a non-GPL build, i.e., excluding M4RI, would be needed if we wanted to use it in `conda`. Am I right, @kalefranz? (BTW: Does anyone know how much of a performance impact excluding M4RI might have?)
And, again considering Miniconda, I'm not sure if we would want to build this without debugging symbols because the `.so` file is a tad large (`~16 MiB`). Stripping that files gives `~1 MiB` -- but usually we don't strip anything in `defaults` and `conda-forge`, I think, so I'm unsure on this.

(cc @conda-forge/core for awareness)

**Comment by kalefranz** [comment]
> Am I right, @kalefranz?

Yup! Exactly. About licensing, splitting out the packages, everything.

**Comment by ocefpaf** [comment]
Thanks for the heads up. Are you planning to on sending PRs to this feedstock or did you already do this on AnacondaRecipes and we should sync it here?

**Comment by mbargull** [comment]
> Are you planning to

Nothing concrete planned, yet. The feedstock isn't even in `AnacondaRecipes` as of now.
I just worked on some performance improvement things in https://github.com/conda/conda/pull/7676 and tested out CryptoMiniSat with `conda`. Some results were quite impressive, so I think it could/would be a huge improvement to be able to use it in `conda`. We would need to address https://github.com/conda-forge/cryptominisat-feedstock/issues/1 and https://github.com/msoos/cryptominisat/issues/513 first, though.
The ping to /core was mostly meant as a "hey, we have some upcoming solver improvements in store, but could need a little help here and there to make it really happen" :wink:.

**Comment by ocefpaf** [comment]
> Some results were quite impressive, so I think it could/would be a huge improvement to be able to use it in conda

Glad to hear that!

> "hey, we have some upcoming solver improvements in store, but could need a little help here and there to make it really happen"

Got it. I'll ping the Windows gurus directly (@gillins, @isuruf, and @patricksnape) for help. I'm OK with the changes you are proposing and I'll see what I can do to make them happen, but I don't think I can help with https://github.com/msoos/cryptominisat/issues/513.

**Comment by msoos** [comment]
Just a note: excluding M4RI has basically no performance impact for the vast majority of CNFs. The kind of  CNF you would be producing here would most likely be completely unaffected. It's used for very weird, mostly hand-crafted, CNF instances.

As I explained on the issue against CMS, there is also a possibility of changing the way you represent the dependency problem you pass to CMS, potentially improving the speed for any SAT solver. I'll help you with that too, if needed, on that issue. Note that everything I will help with is generic and nothing CMS-specific will be used (so you can switch to another SAT solver later if need be).

**Comment by mbargull** [comment]
Does anyone with a little experience in creating split packages and variants want to take a stab at this so it is ready to use once a new CMS version is released? I'm not really familiar with cb3's `outputs` and its intricacies so I'll gladly leave this to someone who is.
What we functionally need for `conda` is basically https://github.com/mbargull/cryptominisat-feedstock/commit/728f5922ca9b5bb3d5f2aa5c102ea046716a15bc (with or without including the `cryptominisat_simple` executable; no strong opinion on this from my side).

**Comment by msoos** [comment]
The new release is now out :) It's 5.6.5 and the startup time has been significantly improved thanks to all your help. There are some other, smallish improvements, notably a CNF dumping API that it a lot nicer.

**Comment by msoos** [comment]
On this note: let me know if you need some improvements regarding the `cryptominsat_simple` thing. I can provide you a compilation flag that doesn't compile that or any executables, creating only the library for example. Just tell me the requirements (if they hit anything on my side) and I can try to help out.

**Comment by msarahan** [comment]
I will take a look at the package split.

**Comment by msarahan** [comment]
@mbargull can you please clarify for me what you think the outputs should be?  One with just the library (no boost dep) and another with the executable (boost dep)?

**Comment by mbargull** [comment]
I think it would make sense to have 3 outputs:
- `libcryptominisat`
  - contains only the library itself,
  - depends only on `zlib`,
- `pycryptosat`:
  - contains only the Python interface,
  - depends only on `python` and `libcryptominisat`,
- `cryptominisat`:
  - contains only the `cryptominisat5` executable,
  - depends on `libcryptominisat` and `boost-cpp` (/`libboost`),
  - maybe also depends on `pycryptosat` to retain backwards compatibility with the current `cryptominisat` packages on `conda-forge`.

There is another executable, `cryptominisat5_simple`, that gets built. This one can go either in `libcryptominisat` or `cryptominisat`. It is about half a megabyte and doesn't depend on `boost-cpp`, so it would be okay to add it to `libcryptominisat`, I guess, but I'm pretty much indifferent about that.

----
Additionally, for `conda` we'd need a non-`GPL` build, i.e., one that does not depend on `m4ri` which can be created by providing `-DMIT=ON`, see https://github.com/mbargull/cryptominisat-feedstock/commit/34ea3d61ebc7f84f6ffa36a5c5b215287141dd2b#diff-44a73bcc045c193c3bd45da87994b03bR19.
Plus, I think it would make sense to build with `-DCMAKE_BUILD_TYPE=Release` (https://github.com/mbargull/cryptominisat-feedstock/commit/34ea3d61ebc7f84f6ffa36a5c5b215287141dd2b#diff-44a73bcc045c193c3bd45da87994b03bR21) since the library gets a little heavy otherwise.

**Comment by msoos** [comment]
I fully support the split a per @mbargull . I think it makes the most sense. As for the `Release` -- beware, this misses all the asserts. Maybe we could have an intermediate compilation target, i.e. no debugging symbols but with asserts. Otherwise, `Release` is appropriate. Maybe I should turn on asserts in `Release` mode... I'll think about it :)

**Comment by mbargull** [comment]
> On this note: let me know if you need some improvements regarding the `cryptominsat_simple` thing. I can provide you a compilation flag that doesn't compile that or any executables, creating only the library for example. Just tell me the requirements (if they hit anything on my side) and I can try to help out.

The current state in gh-11 should work fine, I think. But if (some time in the future) you want to provide separate/independent\* build targets for the library, the simple executable, the non-simple executable, and the python interface, that would surely be nice. (\* independent in the sense that it's possible to build the executables/python extension without needing to rebuild the library)

**Comment by msoos** [comment]
_Re asserts:_ CMS was designed that way and in fact the only asserts that are there should be there always. I have separate debug asserts and that's a very different ballgame. I just don't really understand why cmake Release build runs with `-DNDEBUG`. I am not sure that was a good decision on the part of cmake. So, I don't recommend building Release. Maybe that's a bad decision on my part? I'm a bit confused because running with `-DNDEBUG` was I thought bad practice, but maybe I'm wrong.

_Re build targets:_ Hmmm I just realised that separate targets are already possible. As for "independent" targets... Currently, the way it's done, the executable depends on the library that is compiled during build. It cannot depend on an already installed CMS. To compile e.g. just the library, do: `make libcryptominisat5` instead of plain `make`. For the simple binary + library only: `make cryptominisat5_simple`. For the final binary + library: `make cryptominisat5`. For only the library + the python interface: `make python_interface`. Does this help?

**Comment by msoos** [comment]
_Re asserts:_ Ah, I forgot, I actually personally turn them on for `Release` :D I am removing that now so we won't have this issue. I had to manually change this because it was a really bad user experience. I am also seeing that I managed to break the Windows build you had in the WIP due to turning on `/ZI` for `Release` Windows build. That was my fault. Fixing now that too!

**Comment by msoos** [comment]
Yeah, full rebuilds is the best. Making that cmake thing work has been at least 10-15% of all development time. It's a huge pain. Making it more flexible to save a few compiles is unfortunately not really worth it.

Well, I don't want to replace all asserts with some hand-rolled one... I'll think about it.
